// @vkontakte/vkui/dist/components/ModalRoot/ModalRootContext.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import { createContext, useContext, useRef } from 'react';
import { useIsomorphicLayoutEffect } from "../../lib/useIsomorphicLayoutEffect";
export var ModalRootContext = /*#__PURE__*/createContext({
  updateModalHeight: function updateModalHeight() {
    return undefined;
  },
  registerModal: function registerModal() {
    return undefined;
  },
  isInsideModal: false
});
/**
 * All referenced elements must be static
 */

export function useModalRegistry(id, type) {
  var modalContext = useContext(ModalRootContext);
  var elements = useRef({}).current;
  useIsomorphicLayoutEffect(function () {
    modalContext.registerModal(_objectSpread(_objectSpread({}, elements), {}, {
      type: type,
      id: id
    })); // unset refs on  unmount to prevent leak

    var reset = Object.keys(elements).reduce(function (acc, k) {
      return _objectSpread(_objectSpread({}, acc), {}, _defineProperty({}, k, null));
    }, {
      type: type,
      id: id
    });
    return function () {
      return modalContext.registerModal(reset);
    };
  }, []);
  var refs = useRef({
    modalElement: function modalElement(e) {
      return elements.modalElement = e;
    },
    innerElement: function innerElement(e) {
      return elements.innerElement = e;
    },
    headerElement: function headerElement(e) {
      return elements.headerElement = e;
    },
    contentElement: function contentElement(e) {
      return elements.contentElement = e;
    },
    footerElement: function footerElement(e) {
      return elements.footerElement = e;
    }
  }).current;
  return {
    refs: refs
  };
}
export default ModalRootContext;
