// @vkontakte/vkui/dist/cjs/components/CustomSelect/CustomSelect.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _jsxRuntime = require("../../lib/jsxRuntime");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _SelectMimicry = _interopRequireDefault(require("../SelectMimicry/SelectMimicry"));

var _utils = require("../../lib/utils");

var _classNames = require("../../lib/classNames");

var _CustomScrollView = _interopRequireDefault(require("../CustomScrollView/CustomScrollView"));

var _withAdaptivity = require("../../hoc/withAdaptivity");

var _withPlatform = require("../../hoc/withPlatform");

var _CustomSelectOption = _interopRequireDefault(require("../CustomSelectOption/CustomSelectOption"));

var _getClassName = require("../../helpers/getClassName");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var CustomSelect = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(CustomSelect, _React$Component);

  var _super = _createSuper(CustomSelect);

  function CustomSelect(props) {
    var _this;

    (0, _classCallCheck2.default)(this, CustomSelect);
    _this = _super.call(this, props);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", void 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "keyboardInput", void 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "isControlledOutside", void 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "node", void 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "selectEl", void 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "scrollBoxRef", /*#__PURE__*/(0, _react.createRef)());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "resetKeyboardInput", function () {
      _this.keyboardInput = '';
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getSelectedItem", function () {
      var selectedOptionIndex = _this.state.selectedOptionIndex;
      var options = _this.props.options;

      if (!options.length) {
        return null;
      }

      return options[selectedOptionIndex];
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "open", function () {
      _this.setState(function (_ref) {
        var selectedOptionIndex = _ref.selectedOptionIndex;
        return {
          opened: true,
          focusedOptionIndex: selectedOptionIndex
        };
      }, function () {
        var selectedOptionIndex = _this.state.selectedOptionIndex;

        if (_this.isValidIndex(selectedOptionIndex)) {
          _this.scrollToElement(selectedOptionIndex, true);
        }
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "close", function () {
      _this.resetKeyboardInput();

      _this.setState(function () {
        return {
          opened: false,
          focusedOptionIndex: -1
        };
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "selectFocused", function () {
      var focusedOptionIndex = _this.state.focusedOptionIndex;

      _this.select(focusedOptionIndex);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "select", function (index) {
      var options = _this.props.options;

      if (!_this.isValidIndex(index)) {
        return;
      }

      var item = options[index];

      _this.setState({
        nativeSelectValue: item.value
      }, function () {
        var event = new Event('change', {
          bubbles: true
        });

        _this.selectEl.dispatchEvent(event);
      });

      _this.close();
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onClick", function () {
      _this.state.opened ? _this.close() : _this.open();
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onFocus", function () {
      _this.setState(function () {
        return {
          focused: true
        };
      });

      var event = new Event('focus');

      _this.selectEl.dispatchEvent(event);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onBlur", function () {
      _this.resetKeyboardInput();

      _this.setState(function () {
        return {
          opened: false,
          focused: false
        };
      });

      var event = new Event('blur');

      _this.selectEl.dispatchEvent(event);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleDocumentClick", function (event) {
      var thisNode = _this.node;

      if (_this.state.opened && thisNode && !thisNode.contains(event.target)) {
        _this.onBlur();
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "focusOptionByIndex", function (index) {
      var focusedOptionIndex = _this.state.focusedOptionIndex;
      var options = _this.props.options;
      var oldIndex = focusedOptionIndex;

      if (index < 0) {
        index = options.length - 1;
      } else if (index >= options.length) {
        index = 0;
      }

      if (index === oldIndex) {
        return;
      }

      _this.scrollToElement(index);

      _this.setState(function () {
        return {
          focusedOptionIndex: index
        };
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "focusOption", function (type) {
      var focusedOptionIndex = _this.state.focusedOptionIndex;
      var index = focusedOptionIndex;

      if (type === 'next') {
        index = focusedOptionIndex + 1;
      } else if (type === 'prev') {
        index = focusedOptionIndex - 1;
      }

      _this.focusOptionByIndex(index);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleOptionHover", function (e) {
      _this.setState({
        focusedOptionIndex: Array.prototype.indexOf.call(e.currentTarget.parentNode.children, e.currentTarget)
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleOptionDown", function (e) {
      e.preventDefault();
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "resetFocusedOption", function () {
      _this.setState(function () {
        return {
          focusedOptionIndex: -1
        };
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onKeyboardInput", function (key) {
      var options = _this.props.options;
      var fullInput = _this.keyboardInput + key;
      var optionIndex = options.findIndex(function (option) {
        return option.label.toLowerCase().includes(fullInput);
      });

      if (optionIndex > -1) {
        _this.focusOptionByIndex(optionIndex);
      }

      _this.keyboardInput = fullInput;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onNativeSelectChange", function (e) {
      var value = e.currentTarget.value;

      if (!_this.isControlledOutside) {
        _this.setState({
          selectedOptionIndex: _this.findSelectedIndex(_this.props.options, value)
        });
      }

      if (_this.props.onChange) {
        _this.props.onChange(e);
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleKeyDownSelect", function (event) {
      var opened = _this.state.opened;

      if (event.key.length === 1 && event.key !== ' ') {
        _this.onKeyboardInput(event.key);

        return;
      }

      switch (event.key) {
        case 'ArrowUp':
          event.preventDefault();
          opened ? _this.focusOption('prev') : _this.open();
          break;

        case 'ArrowDown':
          event.preventDefault();
          opened ? _this.focusOption('next') : _this.open();
          break;

        case 'Escape':
          event.preventDefault();

          _this.close();

          break;

        case 'Enter':
        case 'Spacebar':
        case ' ':
          event.preventDefault();
          opened ? _this.selectFocused() : _this.open();
          break;
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleKeyUp", (0, _utils.debounce)(_this.resetKeyboardInput, 1000));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderOption", function (option, index) {
      var _this$state = _this.state,
          focusedOptionIndex = _this$state.focusedOptionIndex,
          selectedOptionIndex = _this$state.selectedOptionIndex;
      var renderOption = _this.props.renderOption;
      var hovered = index === focusedOptionIndex;
      var selected = index === selectedOptionIndex;
      return (0, _jsxRuntime.createScopedElement)(_react.default.Fragment, {
        key: "".concat(option.value)
      }, renderOption({
        option: option,
        hovered: hovered,
        children: option.label,
        selected: selected,
        onClick: _this.selectFocused,
        onMouseDown: _this.handleOptionDown,
        onMouseEnter: _this.handleOptionHover
      }));
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "rootRef", function (element) {
      _this.node = element;
      (0, _utils.setRef)(element, _this.props.getRootRef);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "selectRef", function (element) {
      _this.selectEl = element;
      (0, _utils.setRef)(element, _this.props.getRef);
    });
    var _value = props.value,
        defaultValue = props.defaultValue;
    var initialValue = _value !== undefined ? _value : defaultValue;
    _this.keyboardInput = '';
    _this.state = {
      opened: false,
      focused: false,
      focusedOptionIndex: -1,
      selectedOptionIndex: _this.findSelectedIndex(props.options, initialValue),
      nativeSelectValue: initialValue
    };

    if (props.value !== undefined) {
      _this.isControlledOutside = true;
    }

    return _this;
  }

  (0, _createClass2.default)(CustomSelect, [{
    key: "findSelectedIndex",
    value: function findSelectedIndex(options, value) {
      return options.findIndex(function (item) {
        value = typeof item.value === 'number' ? Number(value) : value;
        return item.value === value;
      });
    }
  }, {
    key: "isValidIndex",
    value: function isValidIndex(index) {
      return index >= 0 && index < this.props.options.length;
    }
  }, {
    key: "scrollToElement",
    value: function scrollToElement(index) {
      var center = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
      var dropdown = this.scrollBoxRef.current;
      var item = dropdown ? dropdown.children[index] : null;

      if (!item) {
        return;
      }

      var dropdownHeight = dropdown.offsetHeight;
      var scrollTop = dropdown.scrollTop;
      var itemTop = item.offsetTop;
      var itemHeight = item.offsetHeight;

      if (center) {
        dropdown.scrollTop = itemTop - dropdownHeight / 2 + itemHeight / 2;
      } else if (itemTop + itemHeight > dropdownHeight + scrollTop) {
        dropdown.scrollTop = itemTop - dropdownHeight + itemHeight;
      } else if (itemTop < scrollTop) {
        dropdown.scrollTop = itemTop;
      }
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      document.addEventListener('click', this.handleDocumentClick, false);
      document.addEventListener('touchend', this.handleDocumentClick, false);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      document.removeEventListener('click', this.handleDocumentClick, false);
      document.removeEventListener('touchend', this.handleDocumentClick, false);
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (prevProps.value !== this.props.value || prevProps.options !== this.props.options) {
        this.isControlledOutside = this.props.value !== undefined;
        var value = this.props.value === undefined ? this.state.nativeSelectValue : this.props.value;
        this.setState({
          nativeSelectValue: value,
          selectedOptionIndex: this.findSelectedIndex(this.props.options, value)
        });
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$state2 = this.state,
          opened = _this$state2.opened,
          nativeSelectValue = _this$state2.nativeSelectValue;
      var _this$props = this.props,
          name = _this$props.name,
          className = _this$props.className,
          getRef = _this$props.getRef,
          getRootRef = _this$props.getRootRef,
          popupDirection = _this$props.popupDirection,
          options = _this$props.options,
          sizeY = _this$props.sizeY,
          platform = _this$props.platform,
          style = _this$props.style,
          onChange = _this$props.onChange,
          onBlur = _this$props.onBlur,
          onFocus = _this$props.onFocus,
          renderOption = _this$props.renderOption,
          children = _this$props.children,
          restProps = (0, _objectWithoutProperties2.default)(_this$props, ["name", "className", "getRef", "getRootRef", "popupDirection", "options", "sizeY", "platform", "style", "onChange", "onBlur", "onFocus", "renderOption", "children"]);
      var selected = this.getSelectedItem();
      var label = selected ? selected.label : undefined;
      return (0, _jsxRuntime.createScopedElement)("label", {
        vkuiClass: (0, _getClassName.getClassName)('CustomSelect', platform),
        className: className,
        style: style,
        ref: this.rootRef
      }, (0, _jsxRuntime.createScopedElement)(_SelectMimicry.default, (0, _extends2.default)({}, restProps, {
        "aria-hidden": true,
        onClick: this.onClick,
        onKeyDown: this.handleKeyDownSelect,
        onKeyUp: this.handleKeyUp,
        onFocus: this.onFocus,
        onBlur: this.onBlur,
        vkuiClass: (0, _classNames.classNames)({
          'CustomSelect__open': opened,
          'CustomSelect__open--popupDirectionTop': popupDirection === 'top'
        }),
        className: className
      }), label), (0, _jsxRuntime.createScopedElement)("select", {
        ref: this.selectRef,
        name: name,
        onChange: this.onNativeSelectChange,
        onBlur: onBlur,
        onFocus: onFocus,
        value: nativeSelectValue,
        vkuiClass: "CustomSelect__control"
      }, options.map(function (item) {
        return (0, _jsxRuntime.createScopedElement)("option", {
          key: "".concat(item.value),
          value: item.value
        });
      })), opened && (0, _jsxRuntime.createScopedElement)("div", {
        vkuiClass: (0, _classNames.classNames)('CustomSelect__options', "CustomSelect__options--sizeY-".concat(sizeY), {
          'CustomSelect__options--popupDirectionTop': popupDirection === 'top'
        }),
        onMouseLeave: this.resetFocusedOption
      }, (0, _jsxRuntime.createScopedElement)(_CustomScrollView.default, {
        boxRef: this.scrollBoxRef
      }, options.map(this.renderOption))));
    }
  }]);
  return CustomSelect;
}(_react.default.Component);

(0, _defineProperty2.default)(CustomSelect, "defaultProps", {
  renderOption: function renderOption(props) {
    return (0, _jsxRuntime.createScopedElement)(_CustomSelectOption.default, props);
  },
  options: []
});

var _default = (0, _withPlatform.withPlatform)((0, _withAdaptivity.withAdaptivity)(CustomSelect, {
  sizeY: true
}));

exports.default = _default;
