// @vkontakte/vkui/dist/cjs/components/RangeSlider/UniversalSlider.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UniversalSlider = void 0;

var _jsxRuntime = require("../../lib/jsxRuntime");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = require("react");

var _Touch = _interopRequireDefault(require("../Touch/Touch"));

var _getClassName = require("../../helpers/getClassName");

var _classNames = require("../../lib/classNames");

var _utils = require("../../lib/utils");

var _math = require("../../helpers/math");

var _withAdaptivity = require("../../hoc/withAdaptivity");

var _withPlatform = require("../../hoc/withPlatform");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var UniversalSliderDumb = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(UniversalSliderDumb, _Component);

  var _super = _createSuper(UniversalSliderDumb);

  function UniversalSliderDumb() {
    var _this;

    (0, _classCallCheck2.default)(this, UniversalSliderDumb);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "dragging", false);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "startX", 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "containerWidth", 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "container", void 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "thumbStart", /*#__PURE__*/(0, _react.createRef)());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "thumbEnd", /*#__PURE__*/(0, _react.createRef)());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onStart", function (e) {
      if (_this.props.disabled) {
        return;
      }

      var boundingRect = _this.container.getBoundingClientRect();

      _this.containerWidth = boundingRect.width;
      var absolutePosition = e.startX - boundingRect.left;

      var value = _this.offsetToValue(absolutePosition);

      _this.dragging = _this.snapDirection(value, e.originalEvent.target);
      _this.startX = absolutePosition;

      _this.props.onChange(_this.updateRange(value), e);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onMove", function (e) {
      if (_this.props.disabled) {
        return;
      }

      var value = _this.offsetToValue(_this.startX + (e.shiftX || 0));

      _this.props.onChange(_this.updateRange(value), e);

      e.originalEvent.preventDefault();
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onEnd", function () {
      if (_this.props.disabled) {
        return;
      }

      _this.dragging = false;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getRef", function (container) {
      _this.container = container;
      (0, _utils.setRef)(container, _this.props.getRootRef);
    });
    return _this;
  }

  (0, _createClass2.default)(UniversalSliderDumb, [{
    key: "updateRange",
    value: function updateRange(value) {
      if (this.props.disabled) {
        return this.props.value;
      }

      var _this$props$value = (0, _slicedToArray2.default)(this.props.value, 2),
          start = _this$props$value[0],
          end = _this$props$value[1];

      if (start == null) {
        return [null, value];
      }

      var dragging = this.dragging;

      if (dragging === 'start') {
        if (value > end) {
          // "перехватиться", если перетянули за конец
          this.dragging = 'end';
          return [end, value];
        }

        return [value, end];
      }

      if (dragging === 'end') {
        if (value < start) {
          // "перехватиться", если перетянули за начало
          this.dragging = 'start';
          return [value, start];
        }

        return [start, value];
      }

      return this.props.value;
    }
  }, {
    key: "offsetToValue",
    value: function offsetToValue(absolute) {
      var _this$props = this.props,
          min = _this$props.min,
          max = _this$props.max,
          step = _this$props.step;
      return (0, _math.rescale)(absolute, [0, this.containerWidth], [min, max], {
        step: step
      });
    }
  }, {
    key: "snapDirection",
    value: function snapDirection(value, target) {
      if (target === this.thumbStart.current) {
        return 'start';
      }

      if (target === this.thumbEnd.current) {
        return 'end';
      }

      var _this$props$value2 = (0, _slicedToArray2.default)(this.props.value, 2),
          start = _this$props$value2[0],
          end = _this$props$value2[1];

      return Math.abs(start - value) <= Math.abs(end - value) ? 'start' : 'end';
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          min = _this$props2.min,
          max = _this$props2.max,
          step = _this$props2.step,
          value = _this$props2.value,
          defaultValue = _this$props2.defaultValue,
          onChange = _this$props2.onChange,
          getRootRef = _this$props2.getRootRef,
          platform = _this$props2.platform,
          sizeY = _this$props2.sizeY,
          disabled = _this$props2.disabled,
          restProps = (0, _objectWithoutProperties2.default)(_this$props2, ["min", "max", "step", "value", "defaultValue", "onChange", "getRootRef", "platform", "sizeY", "disabled"]);

      var toPercent = function toPercent(v) {
        return (v - min) / (max - min) * 100;
      };

      var isRange = value[0] != null;
      var draggerStyle = isRange ? {
        width: "".concat(toPercent(value[1]) - toPercent(value[0]), "%"),
        left: isRange ? "".concat(toPercent(value[0]), "%") : null
      } : {
        width: "".concat(toPercent(value[1]), "%")
      };
      return (0, _jsxRuntime.createScopedElement)(_Touch.default, (0, _extends2.default)({
        "data-value": isRange ? value.join(',') : value
      }, restProps, {
        onStart: this.onStart,
        onMove: this.onMove,
        onEnd: this.onEnd,
        vkuiClass: (0, _classNames.classNames)((0, _getClassName.getClassName)('Slider', platform), "Slider--sizeY-".concat(sizeY), disabled && 'Slider--disabled')
      }), (0, _jsxRuntime.createScopedElement)("div", {
        ref: this.getRef,
        vkuiClass: "Slider__in"
      }, (0, _jsxRuntime.createScopedElement)("div", {
        vkuiClass: "Slider__dragger",
        style: draggerStyle
      }, isRange && (0, _jsxRuntime.createScopedElement)("span", {
        vkuiClass: (0, _classNames.classNames)('Slider__thumb', 'Slider__thumb--start'),
        ref: this.thumbStart
      }), (0, _jsxRuntime.createScopedElement)("span", {
        vkuiClass: (0, _classNames.classNames)('Slider__thumb', 'Slider__thumb--end'),
        ref: this.thumbEnd
      }))));
    }
  }]);
  return UniversalSliderDumb;
}(_react.Component);

var UniversalSlider = (0, _withAdaptivity.withAdaptivity)((0, _withPlatform.withPlatform)(UniversalSliderDumb), {
  sizeY: true
});
exports.UniversalSlider = UniversalSlider;
