// @vkontakte/vkui/dist/components/CustomSelect/CustomSelect.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import { createScopedElement } from "../../lib/jsxRuntime";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import React, { createRef } from 'react';
import SelectMimicry from "../SelectMimicry/SelectMimicry";
import { debounce, setRef } from "../../lib/utils";
import { classNames } from "../../lib/classNames";
import CustomScrollView from "../CustomScrollView/CustomScrollView";
import { withAdaptivity } from "../../hoc/withAdaptivity";
import { withPlatform } from "../../hoc/withPlatform";
import CustomSelectOption from "../CustomSelectOption/CustomSelectOption";
import { getClassName } from "../../helpers/getClassName";

var CustomSelect = /*#__PURE__*/function (_React$Component) {
  _inherits(CustomSelect, _React$Component);

  var _super = _createSuper(CustomSelect);

  function CustomSelect(props) {
    var _this;

    _classCallCheck(this, CustomSelect);

    _this = _super.call(this, props);

    _defineProperty(_assertThisInitialized(_this), "state", void 0);

    _defineProperty(_assertThisInitialized(_this), "keyboardInput", void 0);

    _defineProperty(_assertThisInitialized(_this), "isControlledOutside", void 0);

    _defineProperty(_assertThisInitialized(_this), "node", void 0);

    _defineProperty(_assertThisInitialized(_this), "selectEl", void 0);

    _defineProperty(_assertThisInitialized(_this), "scrollBoxRef", /*#__PURE__*/createRef());

    _defineProperty(_assertThisInitialized(_this), "resetKeyboardInput", function () {
      _this.keyboardInput = '';
    });

    _defineProperty(_assertThisInitialized(_this), "getSelectedItem", function () {
      var selectedOptionIndex = _this.state.selectedOptionIndex;
      var options = _this.props.options;

      if (!options.length) {
        return null;
      }

      return options[selectedOptionIndex];
    });

    _defineProperty(_assertThisInitialized(_this), "open", function () {
      _this.setState(function (_ref) {
        var selectedOptionIndex = _ref.selectedOptionIndex;
        return {
          opened: true,
          focusedOptionIndex: selectedOptionIndex
        };
      }, function () {
        var selectedOptionIndex = _this.state.selectedOptionIndex;

        if (_this.isValidIndex(selectedOptionIndex)) {
          _this.scrollToElement(selectedOptionIndex, true);
        }
      });
    });

    _defineProperty(_assertThisInitialized(_this), "close", function () {
      _this.resetKeyboardInput();

      _this.setState(function () {
        return {
          opened: false,
          focusedOptionIndex: -1
        };
      });
    });

    _defineProperty(_assertThisInitialized(_this), "selectFocused", function () {
      var focusedOptionIndex = _this.state.focusedOptionIndex;

      _this.select(focusedOptionIndex);
    });

    _defineProperty(_assertThisInitialized(_this), "select", function (index) {
      var options = _this.props.options;

      if (!_this.isValidIndex(index)) {
        return;
      }

      var item = options[index];

      _this.setState({
        nativeSelectValue: item.value
      }, function () {
        var event = new Event('change', {
          bubbles: true
        });

        _this.selectEl.dispatchEvent(event);
      });

      _this.close();
    });

    _defineProperty(_assertThisInitialized(_this), "onClick", function () {
      _this.state.opened ? _this.close() : _this.open();
    });

    _defineProperty(_assertThisInitialized(_this), "onFocus", function () {
      _this.setState(function () {
        return {
          focused: true
        };
      });

      var event = new Event('focus');

      _this.selectEl.dispatchEvent(event);
    });

    _defineProperty(_assertThisInitialized(_this), "onBlur", function () {
      _this.resetKeyboardInput();

      _this.setState(function () {
        return {
          opened: false,
          focused: false
        };
      });

      var event = new Event('blur');

      _this.selectEl.dispatchEvent(event);
    });

    _defineProperty(_assertThisInitialized(_this), "handleDocumentClick", function (event) {
      var thisNode = _this.node;

      if (_this.state.opened && thisNode && !thisNode.contains(event.target)) {
        _this.onBlur();
      }
    });

    _defineProperty(_assertThisInitialized(_this), "focusOptionByIndex", function (index) {
      var focusedOptionIndex = _this.state.focusedOptionIndex;
      var options = _this.props.options;
      var oldIndex = focusedOptionIndex;

      if (index < 0) {
        index = options.length - 1;
      } else if (index >= options.length) {
        index = 0;
      }

      if (index === oldIndex) {
        return;
      }

      _this.scrollToElement(index);

      _this.setState(function () {
        return {
          focusedOptionIndex: index
        };
      });
    });

    _defineProperty(_assertThisInitialized(_this), "focusOption", function (type) {
      var focusedOptionIndex = _this.state.focusedOptionIndex;
      var index = focusedOptionIndex;

      if (type === 'next') {
        index = focusedOptionIndex + 1;
      } else if (type === 'prev') {
        index = focusedOptionIndex - 1;
      }

      _this.focusOptionByIndex(index);
    });

    _defineProperty(_assertThisInitialized(_this), "handleOptionHover", function (e) {
      _this.setState({
        focusedOptionIndex: Array.prototype.indexOf.call(e.currentTarget.parentNode.children, e.currentTarget)
      });
    });

    _defineProperty(_assertThisInitialized(_this), "handleOptionDown", function (e) {
      e.preventDefault();
    });

    _defineProperty(_assertThisInitialized(_this), "resetFocusedOption", function () {
      _this.setState(function () {
        return {
          focusedOptionIndex: -1
        };
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onKeyboardInput", function (key) {
      var options = _this.props.options;
      var fullInput = _this.keyboardInput + key;
      var optionIndex = options.findIndex(function (option) {
        return option.label.toLowerCase().includes(fullInput);
      });

      if (optionIndex > -1) {
        _this.focusOptionByIndex(optionIndex);
      }

      _this.keyboardInput = fullInput;
    });

    _defineProperty(_assertThisInitialized(_this), "onNativeSelectChange", function (e) {
      var value = e.currentTarget.value;

      if (!_this.isControlledOutside) {
        _this.setState({
          selectedOptionIndex: _this.findSelectedIndex(_this.props.options, value)
        });
      }

      if (_this.props.onChange) {
        _this.props.onChange(e);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handleKeyDownSelect", function (event) {
      var opened = _this.state.opened;

      if (event.key.length === 1 && event.key !== ' ') {
        _this.onKeyboardInput(event.key);

        return;
      }

      switch (event.key) {
        case 'ArrowUp':
          event.preventDefault();
          opened ? _this.focusOption('prev') : _this.open();
          break;

        case 'ArrowDown':
          event.preventDefault();
          opened ? _this.focusOption('next') : _this.open();
          break;

        case 'Escape':
          event.preventDefault();

          _this.close();

          break;

        case 'Enter':
        case 'Spacebar':
        case ' ':
          event.preventDefault();
          opened ? _this.selectFocused() : _this.open();
          break;
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handleKeyUp", debounce(_this.resetKeyboardInput, 1000));

    _defineProperty(_assertThisInitialized(_this), "renderOption", function (option, index) {
      var _this$state = _this.state,
          focusedOptionIndex = _this$state.focusedOptionIndex,
          selectedOptionIndex = _this$state.selectedOptionIndex;
      var renderOption = _this.props.renderOption;
      var hovered = index === focusedOptionIndex;
      var selected = index === selectedOptionIndex;
      return createScopedElement(React.Fragment, {
        key: "".concat(option.value)
      }, renderOption({
        option: option,
        hovered: hovered,
        children: option.label,
        selected: selected,
        onClick: _this.selectFocused,
        onMouseDown: _this.handleOptionDown,
        onMouseEnter: _this.handleOptionHover
      }));
    });

    _defineProperty(_assertThisInitialized(_this), "rootRef", function (element) {
      _this.node = element;
      setRef(element, _this.props.getRootRef);
    });

    _defineProperty(_assertThisInitialized(_this), "selectRef", function (element) {
      _this.selectEl = element;
      setRef(element, _this.props.getRef);
    });

    var _value = props.value,
        defaultValue = props.defaultValue;
    var initialValue = _value !== undefined ? _value : defaultValue;
    _this.keyboardInput = '';
    _this.state = {
      opened: false,
      focused: false,
      focusedOptionIndex: -1,
      selectedOptionIndex: _this.findSelectedIndex(props.options, initialValue),
      nativeSelectValue: initialValue
    };

    if (props.value !== undefined) {
      _this.isControlledOutside = true;
    }

    return _this;
  }

  _createClass(CustomSelect, [{
    key: "findSelectedIndex",
    value: function findSelectedIndex(options, value) {
      return options.findIndex(function (item) {
        value = typeof item.value === 'number' ? Number(value) : value;
        return item.value === value;
      });
    }
  }, {
    key: "isValidIndex",
    value: function isValidIndex(index) {
      return index >= 0 && index < this.props.options.length;
    }
  }, {
    key: "scrollToElement",
    value: function scrollToElement(index) {
      var center = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
      var dropdown = this.scrollBoxRef.current;
      var item = dropdown ? dropdown.children[index] : null;

      if (!item) {
        return;
      }

      var dropdownHeight = dropdown.offsetHeight;
      var scrollTop = dropdown.scrollTop;
      var itemTop = item.offsetTop;
      var itemHeight = item.offsetHeight;

      if (center) {
        dropdown.scrollTop = itemTop - dropdownHeight / 2 + itemHeight / 2;
      } else if (itemTop + itemHeight > dropdownHeight + scrollTop) {
        dropdown.scrollTop = itemTop - dropdownHeight + itemHeight;
      } else if (itemTop < scrollTop) {
        dropdown.scrollTop = itemTop;
      }
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      document.addEventListener('click', this.handleDocumentClick, false);
      document.addEventListener('touchend', this.handleDocumentClick, false);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      document.removeEventListener('click', this.handleDocumentClick, false);
      document.removeEventListener('touchend', this.handleDocumentClick, false);
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (prevProps.value !== this.props.value || prevProps.options !== this.props.options) {
        this.isControlledOutside = this.props.value !== undefined;
        var value = this.props.value === undefined ? this.state.nativeSelectValue : this.props.value;
        this.setState({
          nativeSelectValue: value,
          selectedOptionIndex: this.findSelectedIndex(this.props.options, value)
        });
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$state2 = this.state,
          opened = _this$state2.opened,
          nativeSelectValue = _this$state2.nativeSelectValue;

      var _this$props = this.props,
          name = _this$props.name,
          className = _this$props.className,
          getRef = _this$props.getRef,
          getRootRef = _this$props.getRootRef,
          popupDirection = _this$props.popupDirection,
          options = _this$props.options,
          sizeY = _this$props.sizeY,
          platform = _this$props.platform,
          style = _this$props.style,
          onChange = _this$props.onChange,
          onBlur = _this$props.onBlur,
          onFocus = _this$props.onFocus,
          renderOption = _this$props.renderOption,
          children = _this$props.children,
          restProps = _objectWithoutProperties(_this$props, ["name", "className", "getRef", "getRootRef", "popupDirection", "options", "sizeY", "platform", "style", "onChange", "onBlur", "onFocus", "renderOption", "children"]);

      var selected = this.getSelectedItem();
      var label = selected ? selected.label : undefined;
      return createScopedElement("label", {
        vkuiClass: getClassName('CustomSelect', platform),
        className: className,
        style: style,
        ref: this.rootRef
      }, createScopedElement(SelectMimicry, _extends({}, restProps, {
        "aria-hidden": true,
        onClick: this.onClick,
        onKeyDown: this.handleKeyDownSelect,
        onKeyUp: this.handleKeyUp,
        onFocus: this.onFocus,
        onBlur: this.onBlur,
        vkuiClass: classNames({
          'CustomSelect__open': opened,
          'CustomSelect__open--popupDirectionTop': popupDirection === 'top'
        }),
        className: className
      }), label), createScopedElement("select", {
        ref: this.selectRef,
        name: name,
        onChange: this.onNativeSelectChange,
        onBlur: onBlur,
        onFocus: onFocus,
        value: nativeSelectValue,
        vkuiClass: "CustomSelect__control"
      }, options.map(function (item) {
        return createScopedElement("option", {
          key: "".concat(item.value),
          value: item.value
        });
      })), opened && createScopedElement("div", {
        vkuiClass: classNames('CustomSelect__options', "CustomSelect__options--sizeY-".concat(sizeY), {
          'CustomSelect__options--popupDirectionTop': popupDirection === 'top'
        }),
        onMouseLeave: this.resetFocusedOption
      }, createScopedElement(CustomScrollView, {
        boxRef: this.scrollBoxRef
      }, options.map(this.renderOption))));
    }
  }]);

  return CustomSelect;
}(React.Component);

_defineProperty(CustomSelect, "defaultProps", {
  renderOption: function renderOption(props) {
    return createScopedElement(CustomSelectOption, props);
  },
  options: []
});

export default withPlatform(withAdaptivity(CustomSelect, {
  sizeY: true
}));
