// @vkontakte/vkui/dist/components/Epic/Epic.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import { createScopedElement } from "../../lib/jsxRuntime";
import React, { useEffect, useRef } from 'react';
import { getClassName } from "../../helpers/getClassName";
import { usePlatform } from "../../hooks/usePlatform";
import { withAdaptivity, ViewWidth } from "../../hoc/withAdaptivity";
import { ScrollSaver } from "./ScrollSaver";
import { getNavId } from "../../lib/getNavId";
export var Epic = function Epic(props) {
  var platform = usePlatform();
  var scroll = useRef({}).current;

  var activeStory = props.activeStory,
      tabbar = props.tabbar,
      children = props.children,
      viewWidth = props.viewWidth,
      restProps = _objectWithoutProperties(props, ["activeStory", "tabbar", "children", "viewWidth"]);

  useEffect(function () {
    if (!tabbar && viewWidth < ViewWidth.SMALL_TABLET) {
      console.warn('[Epic] Using Epic without tabbar is not recommended on mobile');
    }
  }, [viewWidth]);
  var story = React.Children.toArray(children).find(function (story) {
    return getNavId(story.props) === activeStory;
  }) || null;
  return createScopedElement("div", _extends({}, restProps, {
    vkuiClass: getClassName('Epic', platform)
  }), createScopedElement(ScrollSaver, {
    key: activeStory,
    initialScroll: scroll[activeStory] || 0,
    saveScroll: function saveScroll(value) {
      return scroll[activeStory] = value;
    }
  }, story), tabbar);
};
export default withAdaptivity(Epic, {
  viewWidth: true
});
