// @vkontakte/vkui/dist/components/Typography/Title/Title.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import { createScopedElement } from "../../../lib/jsxRuntime";
import { usePlatform } from "../../../hooks/usePlatform";
import { classNames } from "../../../lib/classNames";
import { getClassName } from "../../../helpers/getClassName";
import { ANDROID } from "../../../lib/platform";
import Headline from "../Headline/Headline";

var getComponent = function getComponent(level) {
  if (!level) {
    return 'div';
  }

  return 'h' + level;
};

var getAndroidTitleWeight = function getAndroidTitleWeight(level, weight) {
  if (level === '3') {
    return weight === 'regular' ? weight : 'medium';
  }

  if (level === '2' && weight === 'semibold') {
    return 'medium';
  }

  if (weight === 'heavy') {
    return 'bold';
  }

  return weight;
};

var Title = function Title(_ref) {
  var children = _ref.children,
      weight = _ref.weight,
      level = _ref.level,
      Component = _ref.Component,
      restProps = _objectWithoutProperties(_ref, ["children", "weight", "level", "Component"]);

  var platform = usePlatform();
  var TitleComponent = Component || getComponent(level);
  var titleWeight = platform === ANDROID ? getAndroidTitleWeight(level, weight) : weight;

  if (platform === ANDROID && level === '3') {
    return createScopedElement(Headline, _extends({
      Component: TitleComponent
    }, restProps, {
      weight: titleWeight
    }), children);
  }

  return createScopedElement(TitleComponent, _extends({}, restProps, {
    vkuiClass: classNames(getClassName('Title', platform), "Title--w-".concat(titleWeight), "Title--l-".concat(level))
  }), children);
};

export default Title;
