// @vkontakte/vkui/dist/cjs/components/ModalRoot/ModalRootContext.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useModalRegistry = useModalRegistry;
exports.default = exports.ModalRootContext = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = require("react");

var _useIsomorphicLayoutEffect = require("../../lib/useIsomorphicLayoutEffect");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var ModalRootContext = /*#__PURE__*/(0, _react.createContext)({
  updateModalHeight: function updateModalHeight() {
    return undefined;
  },
  registerModal: function registerModal() {
    return undefined;
  },
  isInsideModal: false
});
/**
 * All referenced elements must be static
 */

exports.ModalRootContext = ModalRootContext;

function useModalRegistry(id, type) {
  var modalContext = (0, _react.useContext)(ModalRootContext);
  var elements = (0, _react.useRef)({}).current;
  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(function () {
    modalContext.registerModal(_objectSpread(_objectSpread({}, elements), {}, {
      type: type,
      id: id
    })); // unset refs on  unmount to prevent leak

    var reset = Object.keys(elements).reduce(function (acc, k) {
      return _objectSpread(_objectSpread({}, acc), {}, (0, _defineProperty2.default)({}, k, null));
    }, {
      type: type,
      id: id
    });
    return function () {
      return modalContext.registerModal(reset);
    };
  }, []);
  var refs = (0, _react.useRef)({
    modalElement: function modalElement(e) {
      return elements.modalElement = e;
    },
    innerElement: function innerElement(e) {
      return elements.innerElement = e;
    },
    headerElement: function headerElement(e) {
      return elements.headerElement = e;
    },
    contentElement: function contentElement(e) {
      return elements.contentElement = e;
    },
    footerElement: function footerElement(e) {
      return elements.footerElement = e;
    }
  }).current;
  return {
    refs: refs
  };
}

var _default = ModalRootContext;
exports.default = _default;
