// @vkontakte/vkui/dist/cjs/components/DatePicker/DatePicker.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _jsxRuntime = require("../../lib/jsxRuntime");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = require("react");

var _Input = _interopRequireDefault(require("../Input/Input"));

var _withAdaptivity = require("../../hoc/withAdaptivity");

var _utils = require("../../lib/utils");

var _CustomSelect = _interopRequireDefault(require("../CustomSelect/CustomSelect"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var DefaultMonths = ['Января', 'Февраля', 'Марта', 'Апреля', 'Мая', 'Июня', 'Июля', 'Августа', 'Сентября', 'Октября', 'Ноября', 'Декабря'];

var DatePicker = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(DatePicker, _Component);

  var _super = _createSuper(DatePicker);

  function DatePicker(props) {
    var _this;

    (0, _classCallCheck2.default)(this, DatePicker);
    _this = _super.call(this, props);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "parseInputDate", function (date) {
      var splited = date.split('-');
      return {
        day: Number(splited[2]),
        month: Number(splited[1]),
        year: Number(splited[0])
      };
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getDaysInMonth", function (year, month) {
      return new Date(year, month, 0).getDate();
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getMonthMaxDay", function () {
      var _this$state = _this.state,
          month = _this$state.month,
          year = _this$state.year;

      if (!month) {
        return 31;
      }

      if (!year) {
        return _this.getDaysInMonth(2016, month);
      }

      return _this.getDaysInMonth(year, month);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getDayOptions", function () {
      var maxMonthDay = _this.getMonthMaxDay();

      var array = new Array(maxMonthDay);

      for (var i = 0; i < maxMonthDay; i++) {
        var _value = i + 1;

        array[i] = {
          label: String(_value),
          value: _value
        };
      }

      return array;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getMonthOptions", function () {
      var monthNames = _this.props.monthNames;
      return (monthNames || DefaultMonths).map(function (name, index) {
        var value = index + 1;
        return {
          label: name,
          value: value
        };
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getYearOptions", function () {
      var _this$props = _this.props,
          max = _this$props.max,
          min = _this$props.min;
      var yearOptions = [];
      var maxYear = max.year;
      var minYear = min.year;

      for (var _value2 = maxYear; _value2 >= minYear; _value2--) {
        yearOptions.push({
          label: String(_value2),
          value: _value2
        });
      }

      return yearOptions;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onSelectChange", function (e) {
      var onDateChange = _this.props.onDateChange;

      _this.setState((0, _defineProperty2.default)({}, e.target.name, Number(e.target.value)), function () {
        onDateChange && onDateChange(_this.state);
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onStringChange", function (e) {
      var onDateChange = _this.props.onDateChange;

      var date = _this.parseInputDate(e.currentTarget.value);

      _this.setState(function () {
        return _objectSpread({}, date);
      });

      onDateChange && onDateChange(date);
    });
    _this.state = props.defaultValue ? props.defaultValue : {
      day: 0,
      month: 0,
      year: 0
    };
    return _this;
  }

  (0, _createClass2.default)(DatePicker, [{
    key: "convertToInputFormat",
    // Переводим state к формату гг-мм-дд
    value: function convertToInputFormat(date) {
      var day = date.day,
          month = date.month,
          year = date.year;
      return "".concat(year, "-").concat((0, _utils.leadingZero)(month), "-").concat((0, _utils.leadingZero)(day));
    }
  }, {
    key: "customView",
    value: function customView() {
      var _this$props2 = this.props,
          name = _this$props2.name,
          min = _this$props2.min,
          max = _this$props2.max,
          dayPlaceholder = _this$props2.dayPlaceholder,
          monthPlaceholder = _this$props2.monthPlaceholder,
          yearPlaceholder = _this$props2.yearPlaceholder,
          popupDirection = _this$props2.popupDirection,
          defaultValue = _this$props2.defaultValue,
          hasMouse = _this$props2.hasMouse,
          monthNames = _this$props2.monthNames,
          restProps = (0, _objectWithoutProperties2.default)(_this$props2, ["name", "min", "max", "dayPlaceholder", "monthPlaceholder", "yearPlaceholder", "popupDirection", "defaultValue", "hasMouse", "monthNames"]);
      var _this$state2 = this.state,
          day = _this$state2.day,
          month = _this$state2.month,
          year = _this$state2.year;
      return (0, _jsxRuntime.createScopedElement)("div", (0, _extends2.default)({
        vkuiClass: "DatePicker"
      }, restProps), (0, _jsxRuntime.createScopedElement)("div", {
        vkuiClass: "DatePicker__container"
      }, (0, _jsxRuntime.createScopedElement)("div", {
        vkuiClass: "DatePicker__day"
      }, (0, _jsxRuntime.createScopedElement)(_CustomSelect.default, {
        name: "day",
        value: day,
        options: this.getDayOptions(),
        placeholder: dayPlaceholder,
        popupDirection: popupDirection,
        onChange: this.onSelectChange
      })), (0, _jsxRuntime.createScopedElement)("div", {
        vkuiClass: "DatePicker__month"
      }, (0, _jsxRuntime.createScopedElement)(_CustomSelect.default, {
        name: "month",
        value: month,
        options: this.getMonthOptions(),
        placeholder: monthPlaceholder,
        popupDirection: popupDirection,
        onChange: this.onSelectChange
      })), (0, _jsxRuntime.createScopedElement)("div", {
        vkuiClass: "DatePicker__year"
      }, (0, _jsxRuntime.createScopedElement)(_CustomSelect.default, {
        name: "year",
        value: year,
        options: this.getYearOptions(),
        placeholder: yearPlaceholder,
        popupDirection: popupDirection,
        onChange: this.onSelectChange
      }))), (0, _jsxRuntime.createScopedElement)("input", {
        type: "hidden",
        name: name,
        value: this.convertToInputFormat(this.state)
      }));
    }
  }, {
    key: "nativeView",
    value: function nativeView() {
      var _this$props3 = this.props,
          name = _this$props3.name,
          min = _this$props3.min,
          max = _this$props3.max,
          dayPlaceholder = _this$props3.dayPlaceholder,
          monthPlaceholder = _this$props3.monthPlaceholder,
          yearPlaceholder = _this$props3.yearPlaceholder,
          popupDirection = _this$props3.popupDirection,
          defaultValue = _this$props3.defaultValue,
          hasMouse = _this$props3.hasMouse,
          restProps = (0, _objectWithoutProperties2.default)(_this$props3, ["name", "min", "max", "dayPlaceholder", "monthPlaceholder", "yearPlaceholder", "popupDirection", "defaultValue", "hasMouse"]);
      var _this$state3 = this.state,
          day = _this$state3.day,
          month = _this$state3.month,
          year = _this$state3.year;

      if (day && month && year) {
        return (0, _jsxRuntime.createScopedElement)(_Input.default, (0, _extends2.default)({}, restProps, {
          name: name,
          type: "date",
          defaultValue: this.convertToInputFormat(this.state),
          onChange: this.onStringChange,
          min: this.convertToInputFormat(min),
          max: this.convertToInputFormat(max)
        }));
      }

      return (0, _jsxRuntime.createScopedElement)(_Input.default, (0, _extends2.default)({}, restProps, {
        name: name,
        type: "date",
        onChange: this.onStringChange,
        min: this.convertToInputFormat(min),
        max: this.convertToInputFormat(max)
      }));
    }
  }, {
    key: "render",
    value: function render() {
      return this.props.hasMouse ? this.customView() : this.nativeView();
    }
  }]);
  return DatePicker;
}(_react.Component);

(0, _defineProperty2.default)(DatePicker, "defaultProps", {
  min: {
    day: 0,
    month: 0,
    year: 0
  },
  max: {
    day: 31,
    month: 12,
    year: 2100
  }
});

var _default = (0, _withAdaptivity.withAdaptivity)(DatePicker, {
  hasMouse: true
});

exports.default = _default;
