// @vkontakte/vkui/dist/cjs/components/Typography/Title/Title.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _jsxRuntime = require("../../../lib/jsxRuntime");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _usePlatform = require("../../../hooks/usePlatform");

var _classNames = require("../../../lib/classNames");

var _getClassName = require("../../../helpers/getClassName");

var _platform = require("../../../lib/platform");

var _Headline = _interopRequireDefault(require("../Headline/Headline"));

var getComponent = function getComponent(level) {
  if (!level) {
    return 'div';
  }

  return 'h' + level;
};

var getAndroidTitleWeight = function getAndroidTitleWeight(level, weight) {
  if (level === '3') {
    return weight === 'regular' ? weight : 'medium';
  }

  if (level === '2' && weight === 'semibold') {
    return 'medium';
  }

  if (weight === 'heavy') {
    return 'bold';
  }

  return weight;
};

var Title = function Title(_ref) {
  var children = _ref.children,
      weight = _ref.weight,
      level = _ref.level,
      Component = _ref.Component,
      restProps = (0, _objectWithoutProperties2.default)(_ref, ["children", "weight", "level", "Component"]);
  var platform = (0, _usePlatform.usePlatform)();
  var TitleComponent = Component || getComponent(level);
  var titleWeight = platform === _platform.ANDROID ? getAndroidTitleWeight(level, weight) : weight;

  if (platform === _platform.ANDROID && level === '3') {
    return (0, _jsxRuntime.createScopedElement)(_Headline.default, (0, _extends2.default)({
      Component: TitleComponent
    }, restProps, {
      weight: titleWeight
    }), children);
  }

  return (0, _jsxRuntime.createScopedElement)(TitleComponent, (0, _extends2.default)({}, restProps, {
    vkuiClass: (0, _classNames.classNames)((0, _getClassName.getClassName)('Title', platform), "Title--w-".concat(titleWeight), "Title--l-".concat(level))
  }), children);
};

var _default = Title;
exports.default = _default;
