// @vkontakte/vkui/dist/cjs/components/Banner/Banner.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _jsxRuntime = require("../../lib/jsxRuntime");

var _react = require("react");

var _getClassName = require("../../helpers/getClassName");

var _classNames = require("../../lib/classNames");

var _usePlatform = require("../../hooks/usePlatform");

var _platform = require("../../lib/platform");

var _icons = require("@vkontakte/icons");

var _Tappable = _interopRequireDefault(require("../Tappable/Tappable"));

var _Headline = _interopRequireDefault(require("../Typography/Headline/Headline"));

var _Caption = _interopRequireDefault(require("../Typography/Caption/Caption"));

var _Text = _interopRequireDefault(require("../Typography/Text/Text"));

var _utils = require("../../lib/utils");

var _Title = _interopRequireDefault(require("../Typography/Title/Title"));

function renderHeader(_ref) {
  var size = _ref.size,
      header = _ref.header;

  switch (size) {
    case 's':
      return (0, _jsxRuntime.createScopedElement)(_Headline.default, {
        weight: "medium",
        vkuiClass: "Banner__header"
      }, header);

    case 'm':
      return (0, _jsxRuntime.createScopedElement)(_Title.default, {
        level: "2",
        weight: "medium",
        vkuiClass: "Banner__header"
      }, header);
  }
}

function renderSubheader(_ref2) {
  var size = _ref2.size,
      subheader = _ref2.subheader;

  switch (size) {
    case 's':
      return (0, _jsxRuntime.createScopedElement)(_Caption.default, {
        level: "1",
        weight: "regular",
        vkuiClass: "Banner__subheader"
      }, subheader);

    case 'm':
      return (0, _jsxRuntime.createScopedElement)(_Text.default, {
        weight: "regular",
        vkuiClass: "Banner__subheader"
      }, subheader);
  }
}

var Banner = function Banner(props) {
  var platform = (0, _usePlatform.usePlatform)();
  var mode = props.mode,
      imageTheme = props.imageTheme,
      size = props.size,
      before = props.before,
      asideMode = props.asideMode,
      header = props.header,
      subheader = props.subheader,
      text = props.text,
      children = props.children,
      background = props.background,
      actions = props.actions,
      onDismiss = props.onDismiss,
      restProps = (0, _objectWithoutProperties2.default)(props, ["mode", "imageTheme", "size", "before", "asideMode", "header", "subheader", "text", "children", "background", "actions", "onDismiss"]);
  var InnerComponent = asideMode === 'expand' ? _Tappable.default : 'div';
  var innerProps = asideMode === 'expand' ? {
    activeMode: platform === _platform.IOS ? 'opacity' : 'background'
  } : {};
  return (0, _jsxRuntime.createScopedElement)("div", (0, _extends2.default)({}, restProps, {
    vkuiClass: (0, _classNames.classNames)((0, _getClassName.getClassName)('Banner', platform), "Banner--md-".concat(mode), "Banner--sz-".concat(size), {
      'Banner--inverted': mode === 'image' && imageTheme === 'dark'
    })
  }), (0, _jsxRuntime.createScopedElement)(InnerComponent, (0, _extends2.default)({
    vkuiClass: "Banner__in"
  }, innerProps), mode === 'image' && background && (0, _jsxRuntime.createScopedElement)("div", {
    vkuiClass: "Banner__bg"
  }, background), before && (0, _jsxRuntime.createScopedElement)("div", {
    vkuiClass: "Banner__before"
  }, before), (0, _jsxRuntime.createScopedElement)("div", {
    vkuiClass: "Banner__content"
  }, (0, _utils.hasReactNode)(header) && renderHeader({
    size: size,
    header: header
  }), (0, _utils.hasReactNode)(subheader) && renderSubheader({
    size: size,
    subheader: subheader
  }), (0, _utils.hasReactNode)(text) && (0, _jsxRuntime.createScopedElement)(_Text.default, {
    weight: "regular",
    vkuiClass: "Banner__text"
  }, text), (0, _utils.hasReactNode)(actions) && _react.Children.count(actions) > 0 && (0, _jsxRuntime.createScopedElement)("div", {
    vkuiClass: "Banner__actions"
  }, actions)), asideMode === 'expand' && (0, _jsxRuntime.createScopedElement)("div", {
    vkuiClass: "Banner__expand"
  }, (0, _jsxRuntime.createScopedElement)(_icons.Icon24Chevron, null)), asideMode === 'dismiss' && (0, _jsxRuntime.createScopedElement)("div", {
    vkuiClass: "Banner__dismiss"
  }, (0, _jsxRuntime.createScopedElement)("div", {
    vkuiClass: "Banner__dismissIcon",
    onClick: onDismiss
  }, (platform === _platform.ANDROID || platform === _platform.VKCOM) && (0, _jsxRuntime.createScopedElement)(_icons.Icon24Cancel, null), platform === _platform.IOS && (mode === 'image' ? (0, _jsxRuntime.createScopedElement)(_icons.Icon24DismissDark, null) : (0, _jsxRuntime.createScopedElement)(_icons.Icon24DismissSubstract, null))))));
};

Banner.defaultProps = {
  mode: 'tint',
  size: 's',
  imageTheme: 'dark'
};
var _default = Banner;
exports.default = _default;
