// @vkontakte/vkui/dist/cjs/components/ConfigProvider/ConfigProviderContext.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ConfigProviderContext = exports.defaultConfigProviderProps = exports.WebviewType = exports.Scheme = exports.Appearance = void 0;

var _react = require("react");

var _vkBridge = _interopRequireDefault(require("@vkontakte/vk-bridge"));

var _platform = require("../../lib/platform");

var Appearance;
exports.Appearance = Appearance;

(function (Appearance) {
  Appearance["DARK"] = "dark";
  Appearance["LIGHT"] = "light";
})(Appearance || (exports.Appearance = Appearance = {}));

var Scheme;
exports.Scheme = Scheme;

(function (Scheme) {
  Scheme["DEPRECATED_CLIENT_LIGHT"] = "client_light";
  Scheme["DEPRECATED_CLIENT_DARK"] = "client_dark";
  Scheme["BRIGHT_LIGHT"] = "bright_light";
  Scheme["SPACE_GRAY"] = "space_gray";
  Scheme["VKCOM"] = "vkcom";
})(Scheme || (exports.Scheme = Scheme = {}));

var WebviewType;
exports.WebviewType = WebviewType;

(function (WebviewType) {
  WebviewType["VKAPPS"] = "vkapps";
  WebviewType["INTERNAL"] = "internal";
})(WebviewType || (exports.WebviewType = WebviewType = {}));

var defaultConfigProviderProps = {
  webviewType: WebviewType.VKAPPS,
  isWebView: _vkBridge.default.isWebView(),
  scheme: Scheme.BRIGHT_LIGHT,
  appearance: Appearance.LIGHT,
  transitionMotionEnabled: true,
  platform: (0, _platform.platform)()
};
exports.defaultConfigProviderProps = defaultConfigProviderProps;
var ConfigProviderContext = /*#__PURE__*/(0, _react.createContext)(defaultConfigProviderProps);
exports.ConfigProviderContext = ConfigProviderContext;
