// @vkontakte/vkui/dist/cjs/components/HorizontalCell/HorizontalCell.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HorizontalCell = void 0;

var _jsxRuntime = require("../../lib/jsxRuntime");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classNames = require("../../lib/classNames");

var _getClassName = require("../../helpers/getClassName");

var _usePlatform = require("../../hooks/usePlatform");

var _utils = require("../../lib/utils");

var _Caption = _interopRequireDefault(require("../Typography/Caption/Caption"));

var _Tappable = _interopRequireDefault(require("../Tappable/Tappable"));

var _Subhead = _interopRequireDefault(require("../Typography/Subhead/Subhead"));

var _Avatar = _interopRequireDefault(require("../Avatar/Avatar"));

var CellTypography = function CellTypography(_ref) {
  var size = _ref.size,
      children = _ref.children,
      restProps = (0, _objectWithoutProperties2.default)(_ref, ["size", "children"]);
  return size === 's' ? (0, _jsxRuntime.createScopedElement)(_Caption.default, (0, _extends2.default)({
    level: "2",
    weight: "regular"
  }, restProps), children) : (0, _jsxRuntime.createScopedElement)(_Subhead.default, (0, _extends2.default)({
    weight: "regular"
  }, restProps), children);
};

var HorizontalCell = function HorizontalCell(props) {
  var className = props.className,
      header = props.header,
      subtitle = props.subtitle,
      size = props.size,
      style = props.style,
      children = props.children,
      getRootRef = props.getRootRef,
      getRef = props.getRef,
      restProps = (0, _objectWithoutProperties2.default)(props, ["className", "header", "subtitle", "size", "style", "children", "getRootRef", "getRef"]);
  var platform = (0, _usePlatform.usePlatform)();
  return (0, _jsxRuntime.createScopedElement)("div", {
    vkuiClass: (0, _classNames.classNames)((0, _getClassName.getClassName)('HorizontalCell', platform), "HorizontalCell--".concat(size)),
    ref: getRootRef,
    style: style,
    className: className
  }, (0, _jsxRuntime.createScopedElement)(_Tappable.default, (0, _extends2.default)({
    getRootRef: getRef
  }, restProps), (0, _utils.hasReactNode)(children) && (0, _jsxRuntime.createScopedElement)("div", {
    vkuiClass: "HorizontalCell__image"
  }, children), (0, _jsxRuntime.createScopedElement)("div", {
    vkuiClass: "HorizontalCell__content"
  }, (0, _utils.hasReactNode)(header) && (0, _jsxRuntime.createScopedElement)(CellTypography, {
    size: size,
    vkuiClass: "HorizontalCell__title"
  }, header), (0, _utils.hasReactNode)(subtitle) && (0, _jsxRuntime.createScopedElement)(_Caption.default, {
    weight: "regular",
    level: "1",
    vkuiClass: "HorizontalCell__subtitle"
  }, subtitle))));
};

exports.HorizontalCell = HorizontalCell;
HorizontalCell.defaultProps = {
  size: 's',
  children: (0, _jsxRuntime.createScopedElement)(_Avatar.default, {
    size: 56
  })
};
