// @vkontakte/vkui/dist/components/RichCell/RichCell.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import { createScopedElement } from "../../lib/jsxRuntime";
import { classNames } from "../../lib/classNames";
import { usePlatform } from "../../hooks/usePlatform";
import { getClassName } from "../../helpers/getClassName";
import Tappable from "../Tappable/Tappable";
import { hasReactNode } from "../../lib/utils";
import Text from "../Typography/Text/Text";
import Caption from "../Typography/Caption/Caption";
import { withAdaptivity } from "../../hoc/withAdaptivity";

var RichCell = function RichCell(_ref) {
  var children = _ref.children,
      text = _ref.text,
      caption = _ref.caption,
      before = _ref.before,
      after = _ref.after,
      bottom = _ref.bottom,
      actions = _ref.actions,
      multiline = _ref.multiline,
      Component = _ref.Component,
      onClick = _ref.onClick,
      sizeY = _ref.sizeY,
      restProps = _objectWithoutProperties(_ref, ["children", "text", "caption", "before", "after", "bottom", "actions", "multiline", "Component", "onClick", "sizeY"]);

  var platform = usePlatform();
  var RootComponent = restProps.disabled ? Component : Tappable;
  Component = restProps.disabled ? undefined : Component;
  var props = restProps;

  if (!restProps.disabled) {
    props.Component = restProps.href ? 'a' : Component;
    props.onClick = onClick;
  }

  return createScopedElement(RootComponent, _extends({}, props, {
    vkuiClass: classNames(getClassName('RichCell', platform), {
      'RichCell--mult': multiline
    }, "RichCell--sizeY-".concat(sizeY))
  }), before, createScopedElement("div", {
    vkuiClass: "RichCell__in"
  }, createScopedElement("div", {
    vkuiClass: "RichCell__top"
  }, after, createScopedElement(Text, {
    weight: "medium",
    vkuiClass: "RichCell__content"
  }, createScopedElement("div", {
    vkuiClass: "RichCell__children"
  }, children), hasReactNode(after) && createScopedElement("div", {
    vkuiClass: "RichCell__after"
  }, after)), hasReactNode(text) && createScopedElement(Text, {
    weight: "regular",
    vkuiClass: "RichCell__text"
  }, text), hasReactNode(caption) && createScopedElement(Caption, {
    level: "1",
    weight: "regular",
    vkuiClass: "RichCell__caption"
  }, caption), (hasReactNode(bottom) || hasReactNode(actions)) && createScopedElement("div", {
    vkuiClass: "RichCell__bottom"
  }, bottom, hasReactNode(actions) && createScopedElement("div", {
    vkuiClass: "RichCell__actions"
  }, actions)))));
};

RichCell.defaultProps = {
  Component: 'div'
};
export default withAdaptivity(RichCell, {
  sizeY: true
});
