// @vkontakte/vkui/dist/components/PopoutRoot/PopoutRoot.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import { createScopedElement } from "../../lib/jsxRuntime";
import { useEffect } from 'react';
import { withAdaptivity, ViewWidth } from "../../hoc/withAdaptivity";
import { AppRootPortal } from "../AppRoot/AppRootPortal";
import { blurActiveElement, useDOM } from "../../lib/dom";
export var PopoutRootComponent = function PopoutRootComponent(props) {
  var popout = props.popout,
      modal = props.modal,
      viewWidth = props.viewWidth,
      children = props.children,
      getRootRef = props.getRootRef,
      restProps = _objectWithoutProperties(props, ["popout", "modal", "viewWidth", "children", "getRootRef"]);

  var _useDOM = useDOM(),
      document = _useDOM.document;

  var isDesktop = viewWidth >= ViewWidth.TABLET;
  useEffect(function () {
    popout && blurActiveElement(document);
  }, [!!popout]);
  return createScopedElement("div", _extends({}, restProps, {
    vkuiClass: "PopoutRoot",
    ref: getRootRef
  }), children, createScopedElement(AppRootPortal, null, !!popout && createScopedElement("div", {
    vkuiClass: isDesktop ? 'PopoutRoot--absolute' : 'PopoutRoot__popout'
  }, popout), !!modal && createScopedElement("div", {
    vkuiClass: "PopoutRoot__modal"
  }, modal)));
};
PopoutRootComponent.displayName = 'PopoutRoot';
export var PopoutRoot = withAdaptivity(PopoutRootComponent, {
  viewWidth: true
});
